File: src/pkg11_tf2_py/pkg11_tf2_py/dynamic_broadcaster.py
#!/usr/bin/env python3
"""Dynamic TF broadcaster in Python: world -> robot circling at 20 Hz."""

import math
import rclpy
from rclpy.node import Node
from tf2_ros import TransformBroadcaster
from geometry_msgs.msg import TransformStamped


def quaternion_from_yaw(yaw: float):
    """Return (x, y, z, w) for a rotation about the z-axis."""
    return (0.0, 0.0, math.sin(yaw / 2.0), math.cos(yaw / 2.0))


class DynamicBroadcaster(Node):
    def __init__(self):
        super().__init__('dynamic_broadcaster_py')
        self.br = TransformBroadcaster(self)
        self.timer = self.create_timer(0.05, self.tick)
        self.t = 0.0
        self.get_logger().info('Publishing dynamic TF world -> robot')

    def tick(self):
        tf = TransformStamped()
        tf.header.stamp = self.get_clock().now().to_msg()
        tf.header.frame_id = 'world'
        tf.child_frame_id = 'robot'

        tf.transform.translation.x = math.cos(self.t)
        tf.transform.translation.y = math.sin(self.t)
        tf.transform.translation.z = 0.0

        qx, qy, qz, qw = quaternion_from_yaw(self.t + math.pi / 2)
        tf.transform.rotation.x = qx
        tf.transform.rotation.y = qy
        tf.transform.rotation.z = qz
        tf.transform.rotation.w = qw

        self.br.sendTransform(tf)
        self.t += 0.05


def main(args=None):
    rclpy.init(args=args)
    node = DynamicBroadcaster()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/pkg10_tf2_cpp/src/dynamic_broadcaster.cpp
// dynamic_broadcaster.cpp
// Publishes a transform that changes over time: the child frame "robot"
// rotates around the origin of "world" on a 1m radius circle.

#include <chrono>
#include <cmath>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_ros/transform_broadcaster.h"
#include "geometry_msgs/msg/transform_stamped.hpp"

using namespace std::chrono_literals;

class DynamicBroadcaster : public rclcpp::Node
{
public:
  DynamicBroadcaster() : Node("dynamic_broadcaster"), t_(0.0)
  {
    broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);
    timer_ = create_wall_timer(50ms,
      std::bind(&DynamicBroadcaster::tick, this));
    RCLCPP_INFO(get_logger(), "Publishing dynamic TF world -> robot");
  }

private:
  void tick()
  {
    geometry_msgs::msg::TransformStamped tf;
    tf.header.stamp = now();
    tf.header.frame_id = "world";
    tf.child_frame_id  = "robot";

    tf.transform.translation.x = std::cos(t_);
    tf.transform.translation.y = std::sin(t_);
    tf.transform.translation.z = 0.0;

    // Yaw rotation = t_ radians, so the robot always "faces forward" along its circle
    tf2::Quaternion q;
    q.setRPY(0, 0, t_ + M_PI_2);

    tf.transform.rotation.x = q.x();
    tf.transform.rotation.y = q.y();
    tf.transform.rotation.z = q.z();
    tf.transform.rotation.w = q.w();

    broadcaster_->sendTransform(tf);
    t_ += 0.05;
  }

  std::unique_ptr<tf2_ros::TransformBroadcaster> broadcaster_;
  rclcpp::TimerBase::SharedPtr timer_;
  double t_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<DynamicBroadcaster>());
  rclcpp::shutdown();
  return 0;
}
